feat(ImageGallery): adds ImageGallery and ImageGalleryImage component for gallery view of photos

abbeyhrt opened the pull request:
Closes #229 

Adds two components, ImageGallery and ImageGalleryImage, and their styles, to create the gallery view that exists in the IBM design language site. This also adds a page to the gatsby theme carbon site that provides a ten photo example of the Image Gallery in use.  The gallery ideally matches 1-1 with the [IBM Design Language Gallery](https://www.ibm.com/design/language/gallery/)

**New**

ImageGallery: image gallery works by creating a portal that contains the gallery view when an image is clicked. 

ImageGalleryImage: this is a how images that will be in a gallery are rendered to the page. Based on whether or not the image is in a dialog, the image is either rendered in a column amongst a collection of images or is blown up into a full page view, allowing the users to click or arrow through all the images on the page. Currently requires the images to be in the static folder to render properly. 

ImageGallery mdx file: provides an example of how the ImageGallery works and what it looks like used in markdown files. 


Example with Now deployment: https://gatsby-theme-c-git-fork-abbeyhrt-229-feature-imade-galle-6f06a5.carbon-design-system.now.sh/components/ImageGallery

vpicone commented:
This isn’t working on mobile for me, have you started taking a look at that?

alisonjoseph commented:
@vpicone yep intended behavior at mobile is to disable it. Looks like its doing something weird though right now.

abbeyhrt commented on packages/example/src/pages/components/ImageGallery.mdx:
```diff
@@ -0,0 +1,144 @@
+
+Click on an image to open the gallery.
+
+<Row>
```
Hey! Could you clarify this a bit for me? Not sure what you mean

abbeyhrt replied:
We definitely can, not sure how weird/complex things would get with it all smooshed together. Want me to work on it and see if that's the way we should go?

abbeyhrt replied:
Good idea! Added that in!

abbeyhrt commented:
I added in the spacing tokens and fixed the mobile weirdness(hopefully). I actually fixed the problem with the right arrow, where it was outside of the clickable area, it doesn't click all the way to the edge of the page but neither does the IDL site, it has the 32px margin that isn't clickable. Definitely can change that if we want to though!

abbeyhrt commented:
@alison would you be able to help me/give pointers on how to make it look like it stays at 12 columns at tablet? The colMd and colLg would still need to be there though i think? That is used for the layout when all the images are displayed on the page, unclicked, not for when it is in the dialog

abbeyhrt commented:
@alisonjoseph I added the focus outline to navigation buttons that we talked about in the code review and also added a check to make sure that only videos and images are clickable and are in the gallery

alisonjoseph commented on packages/gatsby-theme-carbon/src/components/ImageGallery/ImageGalleryImage.module.scss:
```diff
@@ -0,0 +1,33 @@
+  align-self: flex-start;
+  left: $spacing-07;
+  color: $text-04;
+  font-weight: bold;
```
```suggestion
  @include carbon--type-style('productice-heading-02');
```

instead of just font-weight: bold

vpicone commented on packages/gatsby-theme-carbon/src/components/ImageGallery/ImageGallery.module.scss:
```diff
@@ -0,0 +1,169 @@
+    max-width: 16.66666667%;
+  }
+
+  :global(.bx--col-lg-3) {
```
This will redefine the rule to the class globally, we'll want to scope it to this component so we don't mess with other grids.

alisonjoseph commented:
Oh, we should test the dark theme on this page

alisonjoseph commented:
@jeanservaas do you have a chance to review this??
https://gatsby-theme-c-git-fork-abbeyhrt-229-feature-imade-galle-6f06a5.carbon-design-system.now.sh/components/ImageGallery

Should function like https://www.ibm.com/design/language/gallery/ however we did add focus states because those were missing on the live site for a11y.

jeanservaas commented:
looks good! (you guys are going to change the gallery to the dark theme though eventually right? 

couple things:

- When you switch to large image view mode, the image title looks like it's jumping position because it was set to stay in the same place as the site name (e.g. IBM Design System)... but that's assuming the blue band is still there. In the Gatsby theme there's no blue band so the title's 40px lower than it should be.


<img width="758" alt="image" src="https://user-images.githubusercontent.com/43144260/63128962-f42d4180-bf7b-11e9-8b29-fa2c4f1514bb.png">

- also looks like the image is slightly off center.

<img width="586" alt="image" src="https://user-images.githubusercontent.com/43144260/63129155-7453a700-bf7c-11e9-89c2-5b1c3f673e78.png">

alisonjoseph commented:
@jeanservaas yep the page is already in dark mode on IDL staging, just waiting for the component to get merged in. https://design-language-website.netlify.com/gallery/

abbeyhrt commented:
@jeanservaas I moved the title up and fixed the image not being centered, let me know if anything else needs to be changed!

vpicone commented on packages/gatsby-theme-carbon/src/components/ImageGallery/ImageGallery.module.scss:
```diff
@@ -0,0 +1,165 @@
+}
+
+.gallery-row {
+  // position: absolute;
```
Let's take these out

abbeyhrt commented:
@vpicone Awesome, thank you! I updated the ImageGallery so it can use the gatsby markdown images now. I was able to keep the button and figure wrappers, so thankfully we can keep the interactive and a11y stuff they provide. I fixed the problem the arrows were having in firefox as well.

vpicone commented:
@abbeyhrt Wow! Looking great! That blur up is so nice. Last couple of things:

1. It looks like you're doubling up with a margin on both the button and the image.

## IDL 
<img width="756" alt="Screen Shot 2019-08-16 at 6 38 15 PM" src="https://user-images.githubusercontent.com/4078018/63203507-70e01e80-c055-11e9-834c-0ab57a2f2056.png">

## This PR
<img width="788" alt="Screen Shot 2019-08-16 at 6 38 12 PM" src="https://user-images.githubusercontent.com/4078018/63203508-70e01e80-c055-11e9-80c8-54a8ad5d5142.png">

2. Can we add a focus to override the default? Test on firefox as well cause it's getting an outline and a border (this might partially be fixed with the margin thing above)
<img width="397" alt="Screen Shot 2019-08-16 at 6 35 03 PM" src="https://user-images.githubusercontent.com/4078018/63203527-8c4b2980-c055-11e9-9bbb-179d1266e66c.png">

jeanservaas commented:
@abbeyhrt 

please don't kill me, there were multiple alignment issues even on IDL gallery page... I told you to move up the gallery image name... but then the "x" looked weirdly off.

just consulted with pete garvin about it and I think we should follow the below spec — it's OK to keep the title of the gallery image lower I think so that it can properly align with the close 'x.'  Super sorry to lead you in circles!

![image](https://user-images.githubusercontent.com/43144260/63291082-7a9d9800-c288-11e9-9e02-70f132253d07.png)
